comp/dogstatsd/replay/writer.go

```diff
@@ -202,32 +202,37 @@ func (tc *TrafficCaptureWriter) Capture(target io.WriteCloser, d time.Duration,
 		tc.oobPacketPoolManager.SetPassthru(false)
 	}
 
-	done := make(chan struct{})
-	defer close(done)
-	go func() {
-		log.Debugf("Capture will be stopped after %v", d)
-
-		select {
-		case <-time.After(d):
-			tc.StopCapture()
-		case <-done:
+	delay := time.After(d)
+Loop:
+	for {
+		if !tc.accepting {
+			// Drain the channel of any msgs that were enqueued in between the last message getting drained
+			// and the timer going off.
+			// this is racy if multiple writers are in play, however we have a single codepath to
```

Comment:
Small nitpick, but this is racy with any number of writers. A single writer is sufficient for this to never terminate or leave messages in the channel.